components/script/dom/history.rs

```diff
@@ -48,9 +61,61 @@ impl History {
         let _ = self.window.upcast::<GlobalScope>().script_to_constellation_chan().send(msg);
         Ok(())
     }
+
+    // https://html.spec.whatwg.org/multipage/#dom-history-pushstate
+    // https://html.spec.whatwg.org/multipage/#dom-history-replacestate
+    fn push_or_replace_state(&self,
+                             cx: *mut JSContext,
+                             data: HandleValue,
+                             _title: DOMString,
+                             _url: Option<USVString>,
+                             _push_or_replace: PushOrReplace) -> ErrorResult {
+        // Step 1
+        let document = self.window.Document();
+        // Step 2
+        if !document.is_fully_active() {
+            return Err(Error::Security);
+        }
+
+        // TODO: Step 3 Optionally abort these steps
+
+        // TODO: Step 4
+
+        // Step 5
+        let serialized_data = StructuredCloneData::write(cx, data)?;
+
+        // TODO: Steps 6-7 Url Handling
+
+        // TODO: Step 8 Push/Replace session history entry
+
+        // TODO: Step 9 Update current entry to represent a GET request
+
+        // TODO: Step 10 Set document's URL to new URL
+
+        // Step 11
+        let global_scope = self.window.upcast::<GlobalScope>();
+        rooted!(in(cx) let mut state = UndefinedValue());
+        serialized_data.read(&global_scope, state.handle_mut());
+
+        // Step 12
+        self.state.set(state.get());
+
+        // TODO: Step 13 Update Document's latest entry to current entry
+
+        Ok(())
+    }
 }
 
 impl HistoryMethods for History {
+    // https://html.spec.whatwg.org/multipage/#dom-history-state
+    #[allow(unsafe_code)]
+    unsafe fn GetState(&self, _cx: *mut JSContext) -> Fallible<JSVal> {
+        if !self.window.Document().is_fully_active() {
+            return Err(Error::Security);
+        }
+        Ok(self.state.get())
```

Comment:
Make the fn safe then replace this with `Ok(unsafe { self.state.get() })`?

Reply:
Same as above. `state.get()` is a safe method.